examples/multi-middleware.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any
+// person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Example showing how to add multiple middlewares to the same server.
+
+use jsonrpsee::{
+	rpc_params,
+	types::traits::Client,
+	utils::server::middleware,
+	ws_client::WsClientBuilder,
+	ws_server::{RpcModule, WsServerBuilder},
+};
+use std::net::SocketAddr;
+use std::time::Instant;
+
+/// Example middleware to measure call execution time.
+#[derive(Clone)]
+struct Timings;
+
+impl middleware::Middleware for Timings {
+	type Instant = Instant;
+
+	fn on_request(&self) -> Self::Instant {
+		Instant::now()
+	}
+
+	fn on_call(&self, name: &str) {
+		println!("[Timings] They called '{}'", name);
+	}
+
+	fn on_result(&self, name: &str, succeess: bool, started_at: Self::Instant) {
+		println!("[Timings] call={}, worked? {}, duration {:?}", name, succeess, started_at.elapsed());
+	}
+
+	fn on_response(&self, started_at: Self::Instant) {
+		println!("[Timings] Response duration {:?}", started_at.elapsed());
+	}
+}
+
+/// Example middleware to keep a watch on the number of total threads started in the system.
+#[derive(Clone)]
+struct ThreadWatcher;
+impl middleware::Middleware for ThreadWatcher {
+	type Instant = isize;
```

Comment:
Can there be negative threads?

Reply:
Yeah, well, like the subtraction of "end - start" can be negative (these are system threads, so anything can happen while the program executes).

Reply:
Yeah, I guess better safe than sorry, and it's just an example